File: model_1/CS3233_2024_Midterm/interval_scheduling_1_solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cstdio>
#include <climits>

using namespace std;

int to_minutes(const string &time_str) {
    int hours = stoi(time_str.substr(0, 2));
    int minutes = stoi(time_str.substr(3, 2));
    return hours * 60 + minutes;
}

string to_time(int minutes) {
    int h = minutes / 60;
    int m = minutes % 60;
    char buffer[6];
    sprintf(buffer, "%02d:%02d", h, m);
    return string(buffer);
}

int main() {
    int n;
    cin >> n;
    cin.ignore();

    vector<string> lines;
    for (int i = 0; i < n; i++) {
        string line;
        getline(cin, line);
        lines.push_back(line);
    }

    vector<int> s_list, e_list;
    for (int i = 0; i < n; i++) {
        string line = lines[i];
        size_t pos = line.find('-');
        string start_str = line.substr(0, pos - 1);
        start_str.erase(remove(start_str.begin(), start_str.end(), ' '), start_str.end());
        string end_str = line.substr(pos + 1);
        end_str.erase(remove(end_str.begin(), end_str.end(), ' '), end_str.end());
        s_list.push_back(to_minutes(start_str));
        e_list.push_back(to_minutes(end_str));
    }

    int first_start = s_list[0];
    int last_end = e_list[n - 1];
    int total_minutes = last_end - first_start;

    auto feasible = [&](int L) -> bool {
        vector<int> A(n, 0);
        vector<int> B(n, 0);

        B[n - 1] = last_end;
        A[n - 1] = min(s_list[n - 1], last_end - L);
        A[n - 1] = max(16 * 60, A[n - 1]);

        for (int i = n - 2; i >= 1; i--) {
            B[i] = A[i + 1];
            A[i] = min(s_list[i], B[i] - L);
            A[i] = max(16 * 60, A[i]);
        }

        if (n >= 2) {
            B[0] = A[1];
        }
        A[0] = min(s_list[0], B[0] - L);

        if (A[0] != first_start) {
            return false;
        }

        for (int i = 0; i < n; i++) {
            if (B[i] < e_list[i]) {
                return false;
            }
            if (B[i] - A[i] < L) {
                return false;
            }
        }
        return true;
    };

    int low = 0;
    int high = total_minutes;
    int best_L = 0;

    while (low <= high) {
        int mid = (low + high) / 2;
        if (feasible(mid)) {
            best_L = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    vector<int> A(n, 0);
    vector<int> B(n, 0);

    B[n - 1] = last_end;
    A[n - 1] = min(s_list[n - 1], last_end - best_L);
    A[n - 1] = max(16 * 60, A[n - 1]);

    for (int i = n - 2; i >= 1; i--) {
        B[i] = A[i + 1];
        A[i] = min(s_list[i], B[i] - best_L);
        A[i] = max(16 * 60, A[i]);
    }

    if (n >= 2) {
        B[0] = A[1];
    }
    A[0] = min(s_list[0], B[0] - best_L);

    for (int i = 0; i < n; i++) {
        cout << to_time(A[i]) << " - " << to_time(B[i]) << endl;
    }

    return 0;
}

File: model_1/CS3233_2024_Midterm/interval_scheduling_1_solution.py
def main():
    import sys
    data = sys.stdin.read().splitlines()
    n = int(data[0].strip())
    intervals = []
    for i in range(1, n + 1):
        line = data[i].strip()
        parts = line.split(' - ')
        start_str = parts[0].strip()
        end_str = parts[1].strip()
        intervals.append((start_str, end_str))
    
    def to_minutes(time_str):
        h, m = time_str.split(':')
        return int(h) * 60 + int(m)
    
    def to_time(minutes):
        h = minutes // 60
        m = minutes % 60
        return f"{h:02d}:{m:02d}"
    
    starts = []
    ends = []
    for start, end in intervals:
        starts.append(to_minutes(start))
        ends.append(to_minutes(end))
    
    first_start = starts[0]
    last_end = ends[-1]
    
    def feasible(L):
        S = [0] * n
        E = [0] * n
        S[-1] = min(starts[-1], last_end - L)
        E[-1] = max(ends[-1], S[-1] + L)
        if E[-1] != last_end:
            return False
        
        for i in range(n-2, -1, -1):
            S[i] = min(starts[i], S[i+1] - L)
            E[i] = max(ends[i], S[i] + L)
            if E[i] > S[i+1]:
                return False
        
        if S[0] != first_start:
            return False
        
        return True
    
    low = 0
    high = last_end - first_start
    ans_L = 0
    while low <= high:
        mid = (low + high) // 2
        if feasible(mid):
            ans_L = mid
            low = mid + 1
        else:
            high = mid - 1
    
    S_opt = [0] * n
    E_opt = [0] * n
    S_opt[-1] = min(starts[-1], last_end - ans_L)
    E_opt[-1] = max(ends[-1], S_opt[-1] + ans_L)
    for i in range(n-2, -1, -1):
        S_opt[i] = min(starts[i], S_opt[i+1] - ans_L)
        E_opt[i] = max(ends[i], S_opt[i] + ans_L)
    
    output_lines = []
    for i in range(n):
        start_time = to_time(S_opt[i])
        end_time = to_time(E_opt[i])
        output_lines.append(f"{start_time} - {end_time}")
    
    for line in output_lines:
        print(line)

if __name__ == '__main__':
    main()